CMakeLists.txt

```diff
@@ -4,30 +4,36 @@
 cmake_minimum_required(VERSION 3.1)
 
 # Workaround to fix wrong compiler on macos.
-if ((APPLE) AND (NOT CMAKE_C_COMPILER))
+if(APPLE AND NOT CMAKE_C_COMPILER)
     set(CMAKE_C_COMPILER "/usr/bin/cc")
 endif()
+
+# Detect if unicorn is compiled as the top-level project
+set(PROJECT_IS_TOP_LEVEL OFF)
```

Comment:
Why does this matter?

Reply:
Does `STATIC` below override this?

Reply:
Maybe an additional thought: the way that CMake (seems to be) designed is that your `CMakeLists.txt` is "supposed" to describe how to build your code and nothing more.

This is why for example I don't enable `UNICORN_STATIC_MSVCRT` when unicorn is not compiled as the top-level project. When you are part of someone else's project all you want to do is describe the required files, flags and includes for your library and let the top-level project decide the CMAKE_C_FLAGS and whether they want to install your library (probably not).

Reply:
> Maybe an additional thought: the way that CMake (seems to be) designed is that your `CMakeLists.txt` is "supposed" to describe how to build your code and nothing more.
> 
> This is why for example I don't enable `UNICORN_STATIC_MSVCRT` when unicorn is not compiled as the top-level project. When you are part of someone else's project all you want to do is describe the required files, flags and includes for your library and let the top-level project decide the CMAKE_C_FLAGS and whether they want to install your library (probably not).



> If you use unicorn as a submodule you do not want to enable a lot of things that you might enable when you are compiling unicorn.
> 
> Eg:
> 
> ```cmake
> # This will cause PROJECT_IS_TOP_LEVEL to be `OFF`, probably the user doesn't want unicorn's tests, install, fuzzing etc.
> add_subdirectory(unicorn)
> 
> # All you want as a (CMake) user unicorn
> add_executable(foo main.cpp)
> target_link_libraries(foo PRIVATE unicorn)
> ```
> 
> My example template: https://github.com/mrexodia/unicorn_template looks like this on Windows
> 
> ![image](https://user-images.githubusercontent.com/2458265/150020448-706cec0c-85b5-4954-8de9-1806f9252a51.png)

Nice project! I just gave a star ;p

> Maybe an additional thought: the way that CMake (seems to be) designed is that your `CMakeLists.txt` is "supposed" to describe how to build your code and nothing more.
> 
> This is why for example I don't enable `UNICORN_STATIC_MSVCRT` when unicorn is not compiled as the top-level project. When you are part of someone else's project all you want to do is describe the required files, flags and includes for your library and let the top-level project decide the CMAKE_C_FLAGS and whether they want to install your library (probably not).

Yes exactly and make sense. In this case, I think some macro like `UNICORN_HAS_{ARCH}` should also be exposed?

Reply:
> In this case, I think some macro like UNICORN_HAS_{ARCH} should also be exposed?

What do you mean with this? If you mean `target_compile_definitions(unicorn PUBLIC UNICORN_HAS_{ARCH})`, probably it's unnecessary as there won't be any compile errors if the user uses the x86 features of unicorn, it will just error in `uc_open` (I guess?). But it doesn't hurt to add it of course.

Reply:
> > In this case, I think some macro like UNICORN_HAS_{ARCH} should also be exposed?
> 
> What do you mean with this? If you mean `target_compile_definitions(unicorn PUBLIC UNICORN_HAS_{ARCH})`, probably it's unnecessary as there won't be any compile errors if the user uses the x86 features of unicorn, it will just error in `uc_open` (I guess?). But it doesn't hurt to add it of course.

Ah yes, those macros are not accessed in `unicorn.h`. Fine.